File: .skills/openclaw-skills/skills/0xnyk/xint/lib/content_audit.ts
import { mkdirSync, writeFileSync, existsSync } from "fs";
import { join } from "path";
import { loadTokens, getValidToken } from "./oauth";
import { getUserTimeline, type TweetWithEngagement } from "./timeline";
import { grokChat } from "./grok";
import { bold, cyan, yellow, dim } from "./format";

const SKILL_DIR = import.meta.dir;
const EXPORTS_DIR = join(SKILL_DIR, "..", "data", "exports");

function ensureExportsDir(): void {
  if (!existsSync(EXPORTS_DIR)) mkdirSync(EXPORTS_DIR, { recursive: true });
}

function formatTweetsForAudit(tweets: TweetWithEngagement[]): string {
  return tweets.map((t, i) => {
    const pct = (t.engagement_rate * 100).toFixed(2);
    const text = t.text.replace(/https:\/\/t\.co\/\S+/g, "").trim();
    return `[${i + 1}] (${pct}% eng, ${t.metrics.impressions} imp, ${t.metrics.likes} likes)\n${text}`;
  }).join("\n\n");
}

export async function cmdContentAudit(args: string[]): Promise<void> {
  let since = "30d";
  let save = false;
  let pages = 5;

  let i = 0;
  while (i < args.length) {
    const arg = args[i];
    switch (arg) {
      case "--since":
        since = args[++i] || "30d";
        break;
      case "--save":
        save = true;
        break;
      case "--pages":
        pages = Math.min(parseInt(args[++i] || "5"), 10);
        break;
      case "--help":
      case "-h":
        printHelp();
        return;
      default:
        break;
    }
    i++;
  }

  let token: string;
  try {
    token = await getValidToken();
  } catch (e: any) {
    console.error("OAuth required. Run 'xint auth setup' first.");
    console.error(`Error: ${e.message}`);
    process.exit(1);
    return;
  }

  const tokens = loadTokens();
  if (!tokens?.user_id) {
    console.error("Could not determine user ID. Re-run 'xint auth setup'.");
    process.exit(1);
    return;
  }

  console.error(`Fetching your tweets (since ${since})...`);

  const tweets = await getUserTimeline(tokens.user_id, token, {
    since,
    pages,
    exclude: ["replies", "retweets"],
  });

  if (tweets.length < 5) {
    console.error("Not enough tweets for a content audit (need at least 5).");
    return;
  }

  console.error(`Analyzing ${tweets.length} tweets...`);

  // Sort by engagement rate
  const sorted = [...tweets].sort((a, b) => b.engagement_rate - a.engagement_rate);
  const top10 = sorted.slice(0, 10);
  const bottom10 = sorted.slice(-10);

  const topContext = formatTweetsForAudit(top10);
  const bottomContext = formatTweetsForAudit(bottom10);

  const prompt = `Analyze these tweets from my account. The first group are my highest-engagement tweets, the second group are my lowest. What content themes/formats drive the most engagement? What should I post more of? Less of? What hooks work best? Be specific and actionable.`;

  console.error("Sending to Grok for analysis...");

  const response = await grokChat(
    [
      {
        role: "system",
        content: "You are a social media strategist analyzing tweet performance data. Provide specific, actionable recommendations based on the engagement metrics.",
      },
      {
        role: "user",
        content: `${prompt}\n\n--- TOP PERFORMERS ---\n${topContext}\n\n--- LOWEST PERFORMERS ---\n${bottomContext}`,
      },
    ],
    { maxTokens: 2048 },
  );

  console.log();
  console.log(bold("Content Audit") + dim(` (${tweets.length} tweets, since ${since})`) + "\n");
  console.log(response.content);
  console.log();
  console.log(dim(`Model: ${response.model} | Tokens: ${response.usage.total_tokens}`));

  if (save) {
    ensureExportsDir();
    const date = new Date().toISOString().slice(0, 10);
    const path = join(EXPORTS_DIR, `content-audit-${date}.md`);
    let md = `# Content Audit — ${date}\n\n`;
    md += `Tweets analyzed: ${tweets.length} (since ${since})\n\n`;
    md += response.content;
    md += `\n\n---\n\nModel: ${response.model} | Tokens: ${response.usage.total_tokens}\n`;
    writeFileSync(path, md);
    console.error(`\nSaved to ${path}`);
  }
}

function printHelp(): void {
  console.log(`
Usage: xint content-audit [options]

AI-powered analysis of your tweet content performance using Grok.
Identifies what works, what doesn't, and actionable recommendations.

Options:
  --since <dur>    Time period (default: 30d)
  --pages <N>      Pages of tweets to fetch (default: 5, max 10)
  --save           Save audit to data/exports/content-audit-{date}.md

Requires: OAuth + XAI_API_KEY

Examples:
  xint content-audit                 # Audit last 30 days
  xint content-audit --since 90d     # Audit last 90 days
  xint content-audit --save          # Save to file
`);
}

File: .skills/openclaw-skills/skills/0xnyk/xint-rs/src/commands/content_audit.rs
use anyhow::Result;
use colored::Colorize;
use std::fs;

use crate::api::{grok, twitter};
use crate::auth::oauth;
use crate::cli::ContentAuditArgs;
use crate::client::XClient;
use crate::config::Config;
use crate::costs;
use crate::models::{GrokMessage, GrokOpts, Tweet};

pub async fn run(args: &ContentAuditArgs, config: &Config, client: &XClient) -> Result<()> {
    let client_id = config.require_client_id()?;
    let api_key = config.require_xai_key()?;
    let (access_token, tokens) =
        oauth::get_valid_token(client, &config.tokens_path(), client_id).await?;

    eprintln!("Fetching tweets (since {})...", args.since);

    let tweets = twitter::get_user_timeline(
        client,
        &access_token,
        &tokens.user_id,
        100,
        5,
        true,
        true,
        Some(&args.since),
    )
    .await?;

    costs::track_cost(
        &config.costs_path(),
        "timeline",
        &format!("/2/users/{}/tweets", tokens.user_id),
        tweets.len() as u64,
    );

    if tweets.is_empty() {
        println!("No tweets found in the specified time window.");
        return Ok(());
    }

    // Identify top and bottom performers
    let mut sorted: Vec<&Tweet> = tweets.iter().collect();
    sorted.sort_by(|a, b| {
        engagement_rate(b)
            .partial_cmp(&engagement_rate(a))
            .unwrap_or(std::cmp::Ordering::Equal)
    });

    let top_5: Vec<&Tweet> = sorted.iter().take(5).copied().collect();
    let bottom_5: Vec<&Tweet> = sorted.iter().rev().take(5).copied().collect();

    // Build prompt for Grok
    let mut tweet_context = String::new();
    tweet_context.push_str("TOP PERFORMING TWEETS:\n");
    for (i, t) in top_5.iter().enumerate() {
        tweet_context.push_str(&format!(
            "{}. [{:.2}% ER, {} likes, {} impressions]\n\"{}\"\n\n",
            i + 1,
            engagement_rate(t),
            t.metrics.likes,
            t.metrics.impressions,
            truncate(&t.text, 300),
        ));
    }
    tweet_context.push_str("\nLOWEST PERFORMING TWEETS:\n");
    for (i, t) in bottom_5.iter().enumerate() {
        tweet_context.push_str(&format!(
            "{}. [{:.2}% ER, {} likes, {} impressions]\n\"{}\"\n\n",
            i + 1,
            engagement_rate(t),
            t.metrics.likes,
            t.metrics.impressions,
            truncate(&t.text, 300),
        ));
    }

    let total_tweets = tweets.len();
    let avg_er = tweets.iter().map(engagement_rate).sum::<f64>() / total_tweets as f64;
    let avg_likes =
        tweets.iter().map(|t| t.metrics.likes).sum::<u64>() as f64 / total_tweets as f64;

    let system_prompt = format!(
        "You are an expert social media strategist analyzing X/Twitter content performance. \
         The user has {} tweets in the last {} with an average engagement rate of {:.2}% \
         and average {} likes per tweet. Analyze their top and bottom performing content \
         and provide actionable recommendations.",
        total_tweets, args.since, avg_er, avg_likes as u64
    );

    let user_prompt = format!(
        "Analyze my tweet performance and give me:\n\
         1. What patterns make my top tweets successful\n\
         2. Why my bottom tweets underperformed\n\
         3. Content strategy recommendations (3-5 specific, actionable tips)\n\
         4. Suggested content themes based on what works\n\n\
         {}",
        tweet_context
    );

    eprintln!("Analyzing with {}...", args.model);

    let http = reqwest::Client::new();
    let opts = GrokOpts {
        model: args.model.clone(),
        temperature: 0.7,
        max_tokens: 1500,
    };
    let messages = vec![
        GrokMessage {
            role: "system".to_string(),
            content: system_prompt,
        },
        GrokMessage {
            role: "user".to_string(),
            content: user_prompt,
        },
    ];

    let response =
        grok::grok_chat_tracked(&http, api_key, &messages, &opts, Some(&config.costs_path()))
            .await?;

    costs::track_cost(&config.costs_path(), "content_audit", "grok", 0);

    // Display
    println!(
        "\n{}  @{} (last {})\n",
        "Content Audit".bold(),
        tokens.username.cyan(),
        args.since
    );
    println!(
        "  {} tweets analyzed | {:.2}% avg ER | {} avg likes\n",
        total_tweets.to_string().yellow(),
        avg_er,
        (avg_likes as u64).to_string().yellow()
    );
    println!("{}\n", response.content);

    if args.save {
        let exports_dir = config.exports_dir();
        fs::create_dir_all(&exports_dir)?;
        let path = exports_dir.join(format!(
            "content-audit-{}-{}.md",
            tokens.username,
            chrono::Utc::now().format("%Y%m%d-%H%M%S")
        ));
        let report = format!(
            "# Content Audit: @{}\n\n\
             **Date:** {}\n\
             **Period:** last {}\n\
             **Tweets analyzed:** {}\n\
             **Avg engagement rate:** {:.2}%\n\
             **Avg likes:** {}\n\n\
             ---\n\n\
             {}\n",
            tokens.username,
            chrono::Utc::now().format("%Y-%m-%d"),
            args.since,
            total_tweets,
            avg_er,
            avg_likes as u64,
            response.content
        );
        fs::write(&path, &report)?;
        eprintln!("Saved to {}", path.display());
    }

    Ok(())
}

fn engagement_rate(t: &Tweet) -> f64 {
    if t.metrics.impressions == 0 {
        return 0.0;
    }
    let eng = t.metrics.likes + t.metrics.retweets + t.metrics.replies + t.metrics.quotes;
    eng as f64 / t.metrics.impressions as f64 * 100.0
}

fn truncate(s: &str, max: usize) -> String {
    if s.len() <= max {
        s.to_string()
    } else {
        let end = s
            .char_indices()
            .nth(max - 3)
            .map(|(i, _)| i)
            .unwrap_or(s.len());
        format!("{}...", &s[..end])
    }
}
